pkg/controller/issueapi/validate_code.go

```diff
@@ -0,0 +1,148 @@
+		UUID:              project.TrimSpaceAndNonPrintable(request.UUID),
+		ExpiresAt:         now.Add(realm.CodeDuration.Duration),
+		LongExpiresAt:     now.Add(realm.LongCodeDuration.Duration),
+	}
+	if membership != nil {
+		vCode.IssuingUserID = membership.UserID
+	}
+	if authApp != nil {
+		vCode.IssuingAppID = authApp.ID
+	}
+
+	// If this realm requires a date but no date was specified, return an error.
+	if realm.RequireDate && request.SymptomDate == "" && request.TestDate == "" {
+		return nil, &issueResult{
+			obsResult:   observability.ResultError("MISSING_REQUIRED_FIELDS"),
+			httpCode:    http.StatusBadRequest,
+			errorReturn: api.Errorf("missing either test or symptom date").WithCode(api.ErrMissingDate),
+		}
+	}
+
+	// Parse SymptomDate and TestDate
+	var result *issueResult
+	vCode.SymptomDate, result = c.parseDate(request.SymptomDate, int(request.TZOffset), &onsetSettings)
+	if result != nil {
+		return nil, result
+	}
+	vCode.TestDate, result = c.parseDate(request.TestDate, int(request.TZOffset), &testSettings)
+	if result != nil {
+		return nil, result
+	}
+
+	// Verify SMS configuration if phone was provided
+	var smsProvider sms.Provider
+	if request.Phone != "" {
+		smsProvider, err := realm.SMSProvider(c.db)
+		if err != nil {
+			logger.Errorw("failed to get sms provider", "error", err)
+			return nil, &issueResult{
+				obsResult:   observability.ResultError("FAILED_TO_GET_SMS_PROVIDER"),
+				httpCode:    http.StatusInternalServerError,
+				errorReturn: api.Errorf("failed to get sms provider").WithCode(api.ErrInternal),
+			}
+		}
+		if smsProvider == nil {
+			err := fmt.Errorf("phone provided, but no sms provider is configured")
+			return nil, &issueResult{
+				obsResult:   observability.ResultError("FAILED_TO_GET_SMS_PROVIDER"),
+				httpCode:    http.StatusBadRequest,
+				errorReturn: api.Error(err),
+			}
+		}
+	}
+
+	if request.Phone == "" || smsProvider == nil {
+		// If this isn't going to be send via SMS, make the long code expiration time same as short.
+		// This is because the long code will never be shown or sent.
+		vCode.LongExpiresAt = vCode.ExpiresAt
+	}
+
+	// If there is a client-provided UUID, check if a code has already been issued.
+	// this prevents us from consuming quota on conflict.
+	if vCode.UUID != "" {
+		if code, err := realm.FindVerificationCodeByUUID(c.db, request.UUID); err != nil {
+			if !database.IsNotFound(err) {
+				return nil, &issueResult{
+					obsResult:   observability.ResultError("FAILED_TO_CHECK_UUID"),
+					httpCode:    http.StatusInternalServerError,
+					errorReturn: api.Error(err).WithCode(api.ErrInternal),
+				}
+			}
+		} else if code != nil {
+			return nil, &issueResult{
+				obsResult:   observability.ResultError("UUID_CONFLICT"),
+				httpCode:    http.StatusConflict,
+				errorReturn: api.Errorf("code for %s already exists", request.UUID).WithCode(api.ErrUUIDAlreadyExists),
+			}
+		}
+	}
+
+	vCode.Code = "placeholder"
```

Comment:
Is this because you're about to call `Validate` and it'll fail otherwise?

Reply:
Yea otherwise it's 'code too short'. It looks kinda ugly though.
If we go further with pushing validation to the model we might be able to get rid of some of these weirdnesses.